src/sortedarray.h

```diff
@@ -0,0 +1,173 @@
+
+/**
+ * Compare two values in a SortedArray to determine their order.
+ *
+ * @see ArrayListCompareFunc
+ */
+typedef ArrayListCompareFunc SortedArrayCompareFunc;
+
+
+/**
+ * Definition of a @ref SortedArray
+ */
+struct _SortedArray {
+
+    /**
+     * The arraylist to store the data in sorted order.
+     */
+    ArrayList *array;
+
+    /**
+     * The callback used to determine if two values equal.
+     */
+    SortedArrayEqualFunc equ_func;
+
+    /**
+     * The callback use to determine the order of two values.
+     */
+    SortedArrayCompareFunc cmp_func;
+};
+
+/**
+ * Allocate a new SortedArray for use.
+ *
+ * @param length        Indication to the amount of memory that should be 
+ *                      allocated. If 0 is given, then a default is used.
+ * @param equ_func      The function used to determine if two values in the
+ *                      SortedArray equal. This may not be NULL.
+ * @param cmp_func      The function used to determine the relative order of
+ *                      two values in the SortedArray. This may not be NULL.
+ *
+ * @return              A new SortedArray or NULL if it was not possible to
+ *                      allocate one.
+ */
+SortedArray *sortedarray_new(unsigned int length, 
+                             SortedArrayEqualFunc equ_func, 
+                             SortedArrayCompareFunc cmp_func);
+
+/**
+ * Frees a SortedArray from memory.
+ *
+ * @param sortedarray   The SortedArray to free.
+ */
+void sortedarray_free(SortedArray *sortedarray);
+
+/**
+ * Remove a value from a SortedArray at a specified index while maintaining the
+ * sorted property.
+ *
+ * @param sortedarray   The SortedArray to remove a value from.
+ * @param index         The index to remove from the array.
+ */
+void sortedarray_remove(SortedArray *sortedarray, unsigned int index);
+
+/**
+ * Remove a range of entities from a SortedArray while maintaining the sorted 
+ * property.
+ *
+ * @param sortedarray   The SortedArray to remove the range of values from.
+ * @param index         The starting index of the range to remove.
+ * @param length        The length of the range to remove.
+ */
+void sortedarray_remove_range(SortedArray *sortedarray, unsigned int index,
+                              unsigned int length);
+
+/**
+ * Insert a value into a SortedArray while maintaining the sorted property.
+ *
+ * @param sortedarray   The SortedArray to insert into.
+ * @param data          The data to insert.
+ *
+ * @return              Zero on failure, or a non-zero value if successfull.
+ */
+int sortedarray_insert(SortedArray *sortedarray, SortedArrayValue data);
+
+/**
+ * Find the index of a value in a SortedArray.
+ *
+ * @param sortedarray   The SortedArray to find in.
+ * @param data          The value to find.
+ * @return              The index of the value or -1 if the value is not found.
+ */
+int sortedarray_index_of(SortedArray *sortedarray, SortedArrayValue data);
```

Comment:
There's index_of but no apparent API to retrieve a particular element by its index? That seems like a fundamental operation.